// @aws-sdk/client-budgets/dist-es/models/models_0.js
import { SENSITIVE_STRING } from "@aws-sdk/smithy-client";
import { BudgetsServiceException as __BaseException } from "./BudgetsServiceException";
export class AccessDeniedException extends __BaseException {
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        this.name = "AccessDeniedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
        this.Message = opts.Message;
    }
}
export var ThresholdType;
(function (ThresholdType) {
    ThresholdType["ABSOLUTE_VALUE"] = "ABSOLUTE_VALUE";
    ThresholdType["PERCENTAGE"] = "PERCENTAGE";
})(ThresholdType || (ThresholdType = {}));
export var ActionType;
(function (ActionType) {
    ActionType["IAM"] = "APPLY_IAM_POLICY";
    ActionType["SCP"] = "APPLY_SCP_POLICY";
    ActionType["SSM"] = "RUN_SSM_DOCUMENTS";
})(ActionType || (ActionType = {}));
export var ApprovalModel;
(function (ApprovalModel) {
    ApprovalModel["AUTO"] = "AUTOMATIC";
    ApprovalModel["MANUAL"] = "MANUAL";
})(ApprovalModel || (ApprovalModel = {}));
export var ActionSubType;
(function (ActionSubType) {
    ActionSubType["STOP_EC2"] = "STOP_EC2_INSTANCES";
    ActionSubType["STOP_RDS"] = "STOP_RDS_INSTANCES";
})(ActionSubType || (ActionSubType = {}));
export var NotificationType;
(function (NotificationType) {
    NotificationType["ACTUAL"] = "ACTUAL";
    NotificationType["FORECASTED"] = "FORECASTED";
})(NotificationType || (NotificationType = {}));
export var ActionStatus;
(function (ActionStatus) {
    ActionStatus["Execution_Failure"] = "EXECUTION_FAILURE";
    ActionStatus["Execution_In_Progress"] = "EXECUTION_IN_PROGRESS";
    ActionStatus["Execution_Success"] = "EXECUTION_SUCCESS";
    ActionStatus["Pending"] = "PENDING";
    ActionStatus["Reset_Failure"] = "RESET_FAILURE";
    ActionStatus["Reset_In_Progress"] = "RESET_IN_PROGRESS";
    ActionStatus["Reverse_Failure"] = "REVERSE_FAILURE";
    ActionStatus["Reverse_In_Progress"] = "REVERSE_IN_PROGRESS";
    ActionStatus["Reverse_Success"] = "REVERSE_SUCCESS";
    ActionStatus["Standby"] = "STANDBY";
})(ActionStatus || (ActionStatus = {}));
export var SubscriptionType;
(function (SubscriptionType) {
    SubscriptionType["EMAIL"] = "EMAIL";
    SubscriptionType["SNS"] = "SNS";
})(SubscriptionType || (SubscriptionType = {}));
export var EventType;
(function (EventType) {
    EventType["CreateAction"] = "CREATE_ACTION";
    EventType["DeleteAction"] = "DELETE_ACTION";
    EventType["ExecuteAction"] = "EXECUTE_ACTION";
    EventType["System"] = "SYSTEM";
    EventType["UpdateAction"] = "UPDATE_ACTION";
})(EventType || (EventType = {}));
export var AutoAdjustType;
(function (AutoAdjustType) {
    AutoAdjustType["FORECAST"] = "FORECAST";
    AutoAdjustType["HISTORICAL"] = "HISTORICAL";
})(AutoAdjustType || (AutoAdjustType = {}));
export var BudgetType;
(function (BudgetType) {
    BudgetType["Cost"] = "COST";
    BudgetType["RICoverage"] = "RI_COVERAGE";
    BudgetType["RIUtilization"] = "RI_UTILIZATION";
    BudgetType["SPCoverage"] = "SAVINGS_PLANS_COVERAGE";
    BudgetType["SPUtilization"] = "SAVINGS_PLANS_UTILIZATION";
    BudgetType["Usage"] = "USAGE";
})(BudgetType || (BudgetType = {}));
export var TimeUnit;
(function (TimeUnit) {
    TimeUnit["ANNUALLY"] = "ANNUALLY";
    TimeUnit["DAILY"] = "DAILY";
    TimeUnit["MONTHLY"] = "MONTHLY";
    TimeUnit["QUARTERLY"] = "QUARTERLY";
})(TimeUnit || (TimeUnit = {}));
export var ComparisonOperator;
(function (ComparisonOperator) {
    ComparisonOperator["EQUAL_TO"] = "EQUAL_TO";
    ComparisonOperator["GREATER_THAN"] = "GREATER_THAN";
    ComparisonOperator["LESS_THAN"] = "LESS_THAN";
})(ComparisonOperator || (ComparisonOperator = {}));
export var NotificationState;
(function (NotificationState) {
    NotificationState["ALARM"] = "ALARM";
    NotificationState["OK"] = "OK";
})(NotificationState || (NotificationState = {}));
export class CreationLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "CreationLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "CreationLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CreationLimitExceededException.prototype);
        this.Message = opts.Message;
    }
}
export class DuplicateRecordException extends __BaseException {
    constructor(opts) {
        super({
            name: "DuplicateRecordException",
            $fault: "client",
            ...opts,
        });
        this.name = "DuplicateRecordException";
        this.$fault = "client";
        Object.setPrototypeOf(this, DuplicateRecordException.prototype);
        this.Message = opts.Message;
    }
}
export class InternalErrorException extends __BaseException {
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        });
        this.name = "InternalErrorException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalErrorException.prototype);
        this.Message = opts.Message;
    }
}
export class InvalidParameterException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidParameterException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
        this.Message = opts.Message;
    }
}
export class ThrottlingException extends __BaseException {
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        this.name = "ThrottlingException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.Message = opts.Message;
    }
}
export class NotFoundException extends __BaseException {
    constructor(opts) {
        super({
            name: "NotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "NotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NotFoundException.prototype);
        this.Message = opts.Message;
    }
}
export class ResourceLockedException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceLockedException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceLockedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceLockedException.prototype);
        this.Message = opts.Message;
    }
}
export class InvalidNextTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidNextTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidNextTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidNextTokenException.prototype);
        this.Message = opts.Message;
    }
}
export class ExpiredNextTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "ExpiredNextTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "ExpiredNextTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ExpiredNextTokenException.prototype);
        this.Message = opts.Message;
    }
}
export var ExecutionType;
(function (ExecutionType) {
    ExecutionType["ApproveBudgetAction"] = "APPROVE_BUDGET_ACTION";
    ExecutionType["ResetBudgetAction"] = "RESET_BUDGET_ACTION";
    ExecutionType["RetryBudgetAction"] = "RETRY_BUDGET_ACTION";
    ExecutionType["ReverseBudgetAction"] = "REVERSE_BUDGET_ACTION";
})(ExecutionType || (ExecutionType = {}));
export const SubscriberFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Address && { Address: SENSITIVE_STRING }),
});
export const ActionFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const ActionHistoryDetailsFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Action && { Action: ActionFilterSensitiveLog(obj.Action) }),
});
export const ActionHistoryFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.ActionHistoryDetails && {
        ActionHistoryDetails: ActionHistoryDetailsFilterSensitiveLog(obj.ActionHistoryDetails),
    }),
});
export const NotificationWithSubscribersFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const CreateBudgetRequestFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const CreateBudgetActionRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const CreateNotificationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const CreateSubscriberRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscriber && { Subscriber: SubscriberFilterSensitiveLog(obj.Subscriber) }),
});
export const DeleteBudgetActionResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Action && { Action: ActionFilterSensitiveLog(obj.Action) }),
});
export const DeleteSubscriberRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscriber && { Subscriber: SubscriberFilterSensitiveLog(obj.Subscriber) }),
});
export const DescribeBudgetActionResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Action && { Action: ActionFilterSensitiveLog(obj.Action) }),
});
export const DescribeBudgetActionHistoriesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeBudgetActionsForAccountResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeBudgetActionsForBudgetResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
export const DescribeSubscribersForNotificationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const UpdateBudgetActionRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Subscribers && { Subscribers: obj.Subscribers.map((item) => SubscriberFilterSensitiveLog(item)) }),
});
export const UpdateBudgetActionResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.OldAction && { OldAction: ActionFilterSensitiveLog(obj.OldAction) }),
    ...(obj.NewAction && { NewAction: ActionFilterSensitiveLog(obj.NewAction) }),
});
export const UpdateSubscriberRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.OldSubscriber && { OldSubscriber: SubscriberFilterSensitiveLog(obj.OldSubscriber) }),
    ...(obj.NewSubscriber && { NewSubscriber: SubscriberFilterSensitiveLog(obj.NewSubscriber) }),
});
